File: bindings/php/src/KEM.php
<?php
/**
 * libpqc-dyber - Post-Quantum Cryptography Library
 * Copyright (c) 2024-2026 Dyber, Inc.
 * SPDX-License-Identifier: Apache-2.0 OR MIT
 *
 * KEM (Key Encapsulation Mechanism) class.
 */

declare(strict_types=1);

namespace Dyber\PQC;

use RuntimeException;

final class KEM
{
    /** @var \FFI\CData */
    private $handle;
    private string $algorithm;

    /**
     * Create a KEM context for the specified algorithm.
     *
     * @param string $algorithm Algorithm name, e.g. "ML-KEM-768"
     * @throws RuntimeException if the algorithm is not supported
     */
    public function __construct(string $algorithm)
    {
        $this->algorithm = $algorithm;
        $ffi = PQC::ffi();
        $this->handle = $ffi->pqc_kem_new($algorithm);
        if (\FFI::isNull($this->handle)) {
            throw new RuntimeException("Unsupported KEM algorithm: {$algorithm}");
        }
    }

    public function __destruct()
    {
        if (isset($this->handle) && !$this->freed) {
            PQC::ffi()->pqc_kem_free($this->handle);
        }
    }

    private bool $freed = false;

    public function free(): void
    {
        if (!$this->freed) {
            PQC::ffi()->pqc_kem_free($this->handle);
            $this->freed = true;
        }
    }

    public function algorithm(): string
    {
        return $this->algorithm;
    }

    public function publicKeySize(): int
    {
        return PQC::ffi()->pqc_kem_public_key_size($this->handle);
    }

    public function secretKeySize(): int
    {
        return PQC::ffi()->pqc_kem_secret_key_size($this->handle);
    }

    public function ciphertextSize(): int
    {
        return PQC::ffi()->pqc_kem_ciphertext_size($this->handle);
    }

    public function sharedSecretSize(): int
    {
        return PQC::ffi()->pqc_kem_shared_secret_size($this->handle);
    }

    /**
     * Generate a keypair.
     *
     * @return array{publicKey: string, secretKey: string}
     */
    public function keygen(): array
    {
        $ffi = PQC::ffi();
        $pkSize = $this->publicKeySize();
        $skSize = $this->secretKeySize();

        $pk = $ffi->new("uint8_t[{$pkSize}]");
        $sk = $ffi->new("uint8_t[{$skSize}]");

        $rc = $ffi->pqc_kem_keygen($this->handle, $pk, $sk);
        if ($rc !== 0) {
            throw new RuntimeException("KEM keygen failed: " . $ffi->pqc_status_string($rc));
        }

        return [
            'publicKey' => \FFI::string($pk, $pkSize),
            'secretKey' => \FFI::string($sk, $skSize),
        ];
    }

    /**
     * Encapsulate: generate shared secret and ciphertext from a public key.
     *
     * @param string $publicKey
     * @return array{ciphertext: string, sharedSecret: string}
     */
    public function encaps(string $publicKey): array
    {
        $ffi = PQC::ffi();
        $ctSize = $this->ciphertextSize();
        $ssSize = $this->sharedSecretSize();

        $ct = $ffi->new("uint8_t[{$ctSize}]");
        $ss = $ffi->new("uint8_t[{$ssSize}]");

        $pkBuf = $ffi->new("uint8_t[" . strlen($publicKey) . "]");
        \FFI::memcpy($pkBuf, $publicKey, strlen($publicKey));

        $rc = $ffi->pqc_kem_encaps($this->handle, $ct, $ss, $pkBuf);
        if ($rc !== 0) {
            throw new RuntimeException("KEM encaps failed: " . $ffi->pqc_status_string($rc));
        }

        return [
            'ciphertext' => \FFI::string($ct, $ctSize),
            'sharedSecret' => \FFI::string($ss, $ssSize),
        ];
    }

    /**
     * Decapsulate: recover shared secret from ciphertext using secret key.
     *
     * @param string $ciphertext
     * @param string $secretKey
     * @return string The shared secret
     */
    public function decaps(string $ciphertext, string $secretKey): string
    {
        $ffi = PQC::ffi();
        $ssSize = $this->sharedSecretSize();

        $ss = $ffi->new("uint8_t[{$ssSize}]");

        $ctBuf = $ffi->new("uint8_t[" . strlen($ciphertext) . "]");
        \FFI::memcpy($ctBuf, $ciphertext, strlen($ciphertext));

        $skBuf = $ffi->new("uint8_t[" . strlen($secretKey) . "]");
        \FFI::memcpy($skBuf, $secretKey, strlen($secretKey));

        $rc = $ffi->pqc_kem_decaps($this->handle, $ss, $ctBuf, $skBuf);
        if ($rc !== 0) {
            throw new RuntimeException("KEM decaps failed: " . $ffi->pqc_status_string($rc));
        }

        return \FFI::string($ss, $ssSize);
    }
}

File: bindings/csharp/src/KEM.cs
// libpqc-dyber - Post-Quantum Cryptography Library
// Copyright (c) 2024-2026 Dyber, Inc.
// SPDX-License-Identifier: Apache-2.0 OR MIT

using System;
using System.Runtime.InteropServices;

namespace Dyber.PQC
{
    /// <summary>
    /// Key Encapsulation Mechanism (KEM) operations.
    /// </summary>
    public sealed class KEM : IDisposable
    {
        private IntPtr _handle;
        private bool _disposed;

        /// <summary>
        /// Create a KEM context for the specified algorithm.
        /// </summary>
        /// <param name="algorithm">Algorithm name, e.g. "ML-KEM-768".</param>
        /// <exception cref="ArgumentException">Thrown when the algorithm is not supported.</exception>
        public KEM(string algorithm)
        {
            _handle = NativeMethods.pqc_kem_new(algorithm);
            if (_handle == IntPtr.Zero)
                throw new ArgumentException($"Unsupported or disabled KEM algorithm: {algorithm}");
        }

        /// <summary>The algorithm name.</summary>
        public string AlgorithmName
        {
            get
            {
                ThrowIfDisposed();
                IntPtr ptr = NativeMethods.pqc_kem_algorithm(_handle);
                return Marshal.PtrToStringAnsi(ptr) ?? string.Empty;
            }
        }

        /// <summary>Public key size in bytes.</summary>
        public int PublicKeySize
        {
            get { ThrowIfDisposed(); return (int)(uint)NativeMethods.pqc_kem_public_key_size(_handle); }
        }

        /// <summary>Secret key size in bytes.</summary>
        public int SecretKeySize
        {
            get { ThrowIfDisposed(); return (int)(uint)NativeMethods.pqc_kem_secret_key_size(_handle); }
        }

        /// <summary>Ciphertext size in bytes.</summary>
        public int CiphertextSize
        {
            get { ThrowIfDisposed(); return (int)(uint)NativeMethods.pqc_kem_ciphertext_size(_handle); }
        }

        /// <summary>Shared secret size in bytes.</summary>
        public int SharedSecretSize
        {
            get { ThrowIfDisposed(); return (int)(uint)NativeMethods.pqc_kem_shared_secret_size(_handle); }
        }

        /// <summary>NIST security level (1-5).</summary>
        public int SecurityLevel
        {
            get { ThrowIfDisposed(); return NativeMethods.pqc_kem_security_level(_handle); }
        }

        /// <summary>
        /// Generate a new keypair.
        /// </summary>
        /// <returns>A tuple of (publicKey, secretKey) byte arrays.</returns>
        public (byte[] PublicKey, byte[] SecretKey) Keygen()
        {
            ThrowIfDisposed();
            byte[] pk = new byte[PublicKeySize];
            byte[] sk = new byte[SecretKeySize];
            int rc = NativeMethods.pqc_kem_keygen(_handle, pk, sk);
            if (rc != 0)
                throw new InvalidOperationException($"KEM keygen failed: {GetStatusString(rc)}");
            return (pk, sk);
        }

        /// <summary>
        /// Encapsulate: generate a shared secret and ciphertext from a public key.
        /// </summary>
        /// <param name="publicKey">Recipient's public key.</param>
        /// <returns>A tuple of (ciphertext, sharedSecret) byte arrays.</returns>
        public (byte[] Ciphertext, byte[] SharedSecret) Encaps(byte[] publicKey)
        {
            ThrowIfDisposed();
            if (publicKey == null) throw new ArgumentNullException(nameof(publicKey));
            if (publicKey.Length != PublicKeySize)
                throw new ArgumentException($"Public key must be {PublicKeySize} bytes.");

            byte[] ct = new byte[CiphertextSize];
            byte[] ss = new byte[SharedSecretSize];
            int rc = NativeMethods.pqc_kem_encaps(_handle, ct, ss, publicKey);
            if (rc != 0)
                throw new InvalidOperationException($"KEM encaps failed: {GetStatusString(rc)}");
            return (ct, ss);
        }

        /// <summary>
        /// Decapsulate: recover the shared secret from a ciphertext using the secret key.
        /// </summary>
        /// <param name="ciphertext">The ciphertext from encapsulation.</param>
        /// <param name="secretKey">The recipient's secret key.</param>
        /// <returns>The shared secret byte array.</returns>
        public byte[] Decaps(byte[] ciphertext, byte[] secretKey)
        {
            ThrowIfDisposed();
            if (ciphertext == null) throw new ArgumentNullException(nameof(ciphertext));
            if (secretKey == null) throw new ArgumentNullException(nameof(secretKey));
            if (ciphertext.Length != CiphertextSize)
                throw new ArgumentException($"Ciphertext must be {CiphertextSize} bytes.");
            if (secretKey.Length != SecretKeySize)
                throw new ArgumentException($"Secret key must be {SecretKeySize} bytes.");

            byte[] ss = new byte[SharedSecretSize];
            int rc = NativeMethods.pqc_kem_decaps(_handle, ss, ciphertext, secretKey);
            if (rc != 0)
                throw new InvalidOperationException($"KEM decaps failed: {GetStatusString(rc)}");
            return ss;
        }

        private static string GetStatusString(int status)
        {
            IntPtr ptr = NativeMethods.pqc_status_string(status);
            return ptr != IntPtr.Zero ? Marshal.PtrToStringAnsi(ptr) ?? "unknown error" : "unknown error";
        }

        private void ThrowIfDisposed()
        {
            if (_disposed) throw new ObjectDisposedException(nameof(KEM));
        }

        public void Dispose()
        {
            if (!_disposed)
            {
                if (_handle != IntPtr.Zero)
                {
                    NativeMethods.pqc_kem_free(_handle);
                    _handle = IntPtr.Zero;
                }
                _disposed = true;
            }
        }

        ~KEM() => Dispose();
    }
}
